rfcs/20200424-sparse-domain-isolation.md

```diff
@@ -0,0 +1,620 @@
+                         # z should be a trainable and variable-size instance.
+# forward graph
+
+loss = tf.XXX(...)
+
+# backward graph
+opt = tf.train.AdamOptimizer(learning_rate=0.1)
+update_op = opt.minimize(loss)
+```
+
+## Sparse Domain Isolation
+
+### Overview of Design
+![Overview Flow Chart](20200424-sparse-domain-isolation/overview-flow-chart.png)
+
+### Design Considerations
+
+![Architecture of SDI](20200424-sparse-domain-isolation/architecture.png)
+*   The `tf.Tensor` is still suitable for holding iteration weights, because only a small part of weights will be updated in each iterations.
+
+![Expression Weights Hierarchically](20200424-sparse-domain-isolation/expression-weights-hierarchically.png)
+*   Minimize changing on Tensorflow core and try to use native components.
+
+#### Hash Table
+
+Refer to `MutableHashTable` implemented in `tf.lookup`. The reason is as follows:
+
+*   The `MutableHashTable` meets the basic requirements of sparse weights.
+*   The `tf.lookup` is well designed and easy to maintain.
+*   The save/restore from/to checkpoint is already supported.
+*   Compatible with all kind of distributed strategy(parameters server.)
+
+
+### Detailed Design and Implementation
+
+The trainable warp class `dynamic_embdding.TrainableWrap` inherted from `ResourceVariable` will be introduced, and help us reuse all optimizers.
+
+![detail of SDI](20200424-sparse-domain-isolation/optimizers-reuse-scheme.png)
+
+### APIs Overview
+
+* Name Space
+    * `tf.dynamic_embedding` # core version
+    
+* Creating Weights
+    * `tf.dynamic_embedding.Variable`
+    * `tf.dynamic_embedding.get_variable`
+
+* Trainable Wrap
+    * `tf.dynamic_embedding.TrainableWrap`
+    * `tf.dynamic_embedding.embedding_lookup`
+    * `tf.dynamic_embedding.embedding_lookup_sparse`
+
+* High Level APIs(Optional) 
+    * `tf.dynamic_embedding.csr_matmul`
+    * `tf.dynamic_embedding.coo_matmul`
+    
+
+#### Name Space
+
+All of the APIs are implemented in `tf.dynamic_embedding` package.
+
+```python
+import tensorflow as tf # core version
+import tensorflow_recsys as tfrs # SIG version
+
+weights = tf.dynamic_embedding.get_variable(*args, **kwargs)
+```
+
+#### Creating Sparse Weights
+
+We design the `dynamic_embedding.Variable` based on the `tf.lookup.MutableHashTable` and make it support distributed scenario, there is no need to develop any new operators for the hash table. 
+The `dynamic_embedding.Variable` backed up by a group of hashtables is responsible for actually holding the memory resource of sparse weights.
+
+* `tf.dynamic_embedding.Variable`
+    * `export`
+    * `remove`
+    * `upsert` (update if exist else insert) 
+    * `size`
+* `tf.dynamic_embedding.get_variable`
+
+```python
+@tf_export(v1=["dynamic_embedding.Variable"])
+class Variable(object):
+  """
+  A Distributed version of HashTable(reference from lookup_ops.MutableHashTable).
+  It is designed to dynamically store the Sparse Weights(Parameters) of DLRMs.
```

Comment:
Should be "weights of sparse models such as DLRM"

Reply:
ok, I will fix it.